### PR TITLE
bpo-36594: Fix incorrect use of %p in format strings

### PR DESCRIPTION
**Pull request by ZackerySpytz**
In addition, fix some other minor violations of C99.

https://bugs.python.org/issue36594

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
What is incorrect in the use of %p?

**Comment by ZackerySpytz** [comment]
If the argument for ``%p`` is not a pointer to void, the behavior is undefined.

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
@serhiy-storchaka did you want to give this a review or should I just go ahead and merge it?

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -1244,14 +1244,14 @@ _PyUnicode_Dump(PyObject *op)
...
-    printf("wstr=%p", ascii->wstr);
+    printf("wstr=%p", (void *)ascii->wstr);
 
     if (!(ascii->state.ascii == 1 && ascii->state.compact == 1)) {
         printf(" (%" PY_FORMAT_SIZE_T "u), ", compact->wstr_length);
         if (!ascii->state.compact && compact->utf8 == unicode->data.any)
             printf("shared ");
         printf("utf8=%p (%" PY_FORMAT_SIZE_T "u)",
-               compact->utf8, compact->utf8_length);
+               (void *)compact->utf8, compact->utf8_length);
```
Comment (serhiy-storchaka) [inline Objects/unicodeobject.c:1254]: Is this needed? `void*` and `char*` are compatible.

**Review by serhiy-storchaka — APPROVED** [review]

File: Modules/faulthandler.c [hunk 2]
```diff
@@ -1370,7 +1370,7 @@ void _PyFaulthandler_Fini(void)
 #ifdef HAVE_SIGALTSTACK
     if (stack.ss_sp != NULL) {
         /* Fetch the current alt stack */
-        stack_t current_stack = {};
+        stack_t current_stack = {0};
```
Comment (mdickinson) [inline Modules/faulthandler.c]: Do we know that this is always valid? I'm not sure what we can safely rely on with respect to the `stack_t` structure, but if the first field of that struct isn't a scalar, I'd expect initializing it with `0` to be potentially problematic.

**Review by mdickinson — APPROVED** [review]
LGTM

File: Objects/unicodeobject.c (hunk 1 above)
Comment (ZackerySpytz) [reply Objects/unicodeobject.c:1254]: `void*` and `char*` are different types.

File: Modules/faulthandler.c (hunk 2 above)
Comment (ZackerySpytz) [reply Modules/faulthandler.c]: This code was recently modified to use `memset()` (in b84cb70880a0acfcbbaca7bcda405af08f94d269), so faulthandler.c is no longer part of this PR.

**Comment by brettcannon** [comment]
Thanks!

**Pull request merged**